src/engines/evaluate-join-potential.ts

```diff
@@ -0,0 +1,71 @@
+ *   references for the variables in the network, and must be distinct from the
+ *   head variables of the joint being constructed.
+ * @returns: The potential function for the join which satisfies the conditions
+ *   that for every combination of parents, the potentials sum to unity.  The
+ *   case when there are no parents (unconditioned joint distribution), corresponds
+ *   to the trivial case where there is only a single (empty) combination of
+ *   parent values.
+ */
+export function evaluateJoinPotentials (nodes: FastNode[], cliques: FastClique[], connectedComponents: number[][], separators: number[][], formulas: Formula[], potentials: (FastPotential | null)[], headVariables: number[], parentVariables: number[]): { potentials: FastPotential; supplementalFormulas: Formula[]; supplementalPotentials: (FastPotential | null)[]} {
+  // We start by propagating the join messages to construct a formula for the new
+  // join distribution.  This may create supplementary formulas that were not
+  // encountered in the original message passing to make the cliques consistent.
+  const { joinFormulaId, supplementalFormulas } = propagateJoinMessages(nodes, cliques, connectedComponents, formulas, separators, headVariables, parentVariables)
+  // We initialize an array of potential to contain any new computations and seed it
+  // with the previous the previously computed potentials, and enough new elements to cover
+  // the formulas added during message passing.   Then we evaluate the joint formula using
+  // that context.
+  const amendedPotentials: (FastPotential | null)[] = [...potentials, ...Array(supplementalFormulas.length).fill(null)]
+  const amendedFormulas = [...formulas, ...supplementalFormulas]
+
+  // Now we can evaluate the formula for the joint distribution.  This distribution
+  // will be equivalent to the requested join up to a permutation of the variables
+  // in the domain.
+  const joinPotentials = evaluate(joinFormulaId, nodes, amendedFormulas, amendedPotentials)
+  const joinFormula = amendedFormulas[joinFormulaId]
+  const joinDomain = [...headVariables, ...parentVariables]
+
+  // if any new potentials were computed, then update the cache.
+  amendedPotentials.slice(0, potentials.length).forEach((ps, i) => {
+    if (potentials[i] == null) potentials[i] = ps
+  })
+  // If required, we permute the elements in the potential function to agree with the
+  // requested order of variables in the domain and normalize the potentials to ensure
+  // that they sum to unity.
+  let result: FastPotential = []
+  if (joinFormula.domain.every((n, i) => n === joinDomain[i])) {
+    const total = sum(joinPotentials)
+    result = joinPotentials.map(p => total !== 0 ? p / total : p)
+  } else {
+    result = evaluateMarginalPure(joinPotentials, joinFormula.domain, joinFormula.domain.map(i => nodes[i].levels.length), joinDomain, joinDomain.map(i => nodes[i].levels.length), joinPotentials.length, true)
```

Comment:
Line length and wrapping are questionable here too. While I know it doesn't impact functionality, it does impact readability.

Reply:
I would argue that line length is not the sole criteria for readability.
I could get the shortest line length by creating local definitions:

```
const getLevels = (i) => nodes[i].levels.length
const formulaDomain = joinFormula.domain
const formulaNumberOfLevels = formulaDomain.map( getLevels )
const joinNumberOfLevels = joinDomain.map(getLevels)
result = evaulateMarginalPure( joinPotentials, formulaDomain joinNumbersOfLevels,  joinDomain, joinNumberOfLevels, ... )
```

However this puts an emphasis on each individual argument, rather than the overall flow.    I think we should discuss how to strike a balance for readability that makes sense for the team.